### PR TITLE
implement js-ipfs to remove ipfs daemon dependency

### LINKED ISSUE #53: Implement js-ipfs to remove ipfs daemon dependency
Simplifying the number of external dependencies for mine.js will make it easier for people to get up and running. We can implement js-ipfs to remove the dependency on the ipfs daemon.

### PR DESCRIPTION
**Pull request by gavinuhma**
closes #53

### CODE REVIEW

File: bin/cli.js [hunk 1]
```diff
@@ -4,7 +4,6 @@ const program = require('commander')
 const app = require('../mine.js')
 const pckg = require('../package.json')
 const Web3 = require('web3')
-const ipfsAPI = require('ipfs-api')
```
Comment (anoff) [inline bin/cli.js:7]: Please remove this from `package.json` as well if it's not required anymore

File: lib/ipfs.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ipfs.connect = function(cb) {
+  const node = new IPFS()
+  //node.log.enabled = true
+
+  node.on('init', () => {})
+  node.on('ready', () => {})
+  node.on('stop', () => {})
+
+  node.on('start', () => {
+    console.log('Connected to IPFS. Online:', node.isOnline())
```
Comment (anoff) [inline lib/ipfs.js]: needs emojis :)

File: lib/ipfs.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+const IPFS = require('ipfs')
+const ipfs = {}
+
+ipfs.connect = function(cb) {
+  const node = new IPFS()
+  //node.log.enabled = true
+
+  node.on('init', () => {})
+  node.on('ready', () => {})
```
Comment (anoff) [inline lib/ipfs.js]: if you bind to the events please handle them as well. Otherwise the lines are unnecessary

File: lib/ipfs.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  node.on('stop', () => {})
+
+  node.on('start', () => {
+    console.log('Connected to IPFS. Online:', node.isOnline())
+    cb(null, node)
+  })
+
+  node.on('error', (err) => {
+    console.log('ipfs error', err)
```
Comment (anoff) [inline lib/ipfs.js]: `console.error`

File: lib/ipfs.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Download and upload to storage. Currently implemented with ipfs.
+ */
+
+const IPFS = require('ipfs')
+const ipfs = {}
+
+ipfs.connect = function(cb) {
```
Comment (anoff) [inline lib/ipfs.js]: please use promise based APIs

File: mine.js [hunk 6]
```diff
@@ -10,17 +10,29 @@ const tmp = require('tmp')
...
+  })
+}
+
+async function trainModels (mineAddress, contractAddress, web3, ipfs) {
   const sonar = new Sonar(web3, contractAddress, mineAddress)
 
   console.log(`🔎️  Looking for models to train at ${contractAddress} for mine ${mineAddress}`)
   const modelCount = await sonar.getNumModels()
-  console.log(`💃  ${modelCount} models found`)
+  console.log(` 💃  ${modelCount} models found`)
```
Comment (anoff) [inline mine.js]: no unrelated changes in PRs. Do you have issues with the emoji formatting on your CLI? It works for me

File: mine.js [hunk 7]
```diff
@@ -51,23 +63,28 @@ async function checkForModels (mineAddress, contractAddress, web3, ipfs) {
...
       stdio: config.debug ? 'inherit' : ['ignore', 'ignore', process.stderr]
     }
     const trainStart = new Date()
     const sp = spawn(`syft_cmd generate_gradient`, [`-model ${tmpPaths.model}`, `-input_data ${path.join(__dirname, 'data/adapters/diabetes/diabetes_input.csv')}`, `-target_data ${path.join(__dirname, 'data/adapters/diabetes/diabetes_output.csv')}`, `-gradient ${tmpPaths.gradient}`], childOpts)
-    await new Promise((resolve, reject) => {
-      sp.on('close', code => {
-        if (code) reject(new Error(`error while calling syft, code=${code}`))
-        resolve()
+
+    try {
```
Comment (anoff) [inline mine.js]: no unrelated changes

**Review by anoff — CHANGES_REQUESTED** [review]
Love the idea and amazing depth of contribution! However please try to keep changes to the issue you're trying to solve and stay with the existing code style (promise based, emojis in logs.. )

**Comment by gavinuhma** [comment]
thanks for the quick review. I would have said the same :) I’ll clean it up!

**Comment by anoff** [comment]
Thanks for the fix. Promises look good. However the build is broken due to code style errors: https://travis-ci.org/OpenMined/mine.js/builds/293144535?utm_source=github_status&utm_medium=notification

Check locally by running `npm test`

File: lib/ipfs.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Manages the IPFS connection.
+ */
+
+const IPFS = require('ipfs')
+const ipfs = {}
+
+ipfs.connect = async function() {
```
Comment (anoff) [inline lib/ipfs.js]: Please use named functions on high level APIs. Also async is unnecessary here

```
function connect () {
...
module.exports = {
 connect
}
```

At the moment there is only one function in the file but if someone adds another one to the module, anonymous functions make stacktraces hard to read.

Comment (gavinuhma) [reply lib/ipfs.js]: 👍

**Review by anoff — APPROVED** [review]

**Comment by anoff** [comment]
Looks good 👍 
Let's ignore the code-cov for now 😛 

verified it locally
